Source/NSubstitute/Core/GetCallSpec.cs

```diff
@@ -1,26 +1,45 @@
-﻿namespace NSubstitute.Core
+﻿using System;
+
+namespace NSubstitute.Core
 {
     public class GetCallSpec : IGetCallSpec
     {
-        private readonly ICallStack _callStack;
+        private readonly ICallCollection _callCollection;
         private readonly IPendingSpecification _pendingSpecification;
         private readonly ICallSpecificationFactory _callSpecificationFactory;
         private readonly ICallActions _callActions;
 
-        public GetCallSpec(ICallStack callStack, IPendingSpecification pendingSpecification,
+        public GetCallSpec(ICallCollection callCollection, IPendingSpecification pendingSpecification,
             ICallSpecificationFactory callSpecificationFactory, ICallActions callActions)
         {
-            _callStack = callStack;
+            _callCollection = callCollection;
             _pendingSpecification = pendingSpecification;
             _callSpecificationFactory = callSpecificationFactory;
             _callActions = callActions;
         }
 
-        public ICallSpecification FromLastCall(MatchArgs matchArgs)
+        public ICallSpecification FromPendingSpecification(MatchArgs matchArgs)
         {
-            return _pendingSpecification.HasPendingCallSpec()
-                               ? FromExistingSpec(_pendingSpecification.UseCallSpec(), matchArgs)
-                               : FromCall(_callStack.Pop(), matchArgs);
+            if (!_pendingSpecification.HasPendingCallSpecInfo())
+            {
+                throw new InvalidOperationException("No pending specification or previous call info.");
```

Comment:
No, these are just guards to ensure that developer didn't make a mistake. They shouldn't never happen during normal usage. However, I wasn't comfortable to just keep potential places for NullReferenceExceptions and decided to throw something meaningful.

I've already followed the approach you described and introduced [CouldNotSetReturnDueToMissingInfoAboutPreviousCallException](https://github.com/nsubstitute/NSubstitute/blob/84c6a81dddbdea9fd580a89d979ea42981fb40d5/Source/NSubstitute/Exceptions/CouldNotSetReturnException.cs#L48) for places that potentially could happen.